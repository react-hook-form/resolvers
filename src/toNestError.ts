import {
  set,
  get,
  FieldErrors,
  Field,
  ResolverOptions,
  FieldValues,
} from 'react-hook-form';
import { validateFieldsNatively } from './validateFieldsNatively';

export const toNestError = <TFieldValues extends FieldValues>(
  errors: FieldErrors,
  options: ResolverOptions<TFieldValues>,
): FieldErrors<TFieldValues> => {
  options.shouldUseNativeValidation && validateFieldsNatively(errors, options);

  const fieldErrors = {} as FieldErrors<TFieldValues>;
  for (const path in errors) {
    const field = get(options.fields, path) as Field['_f'] | undefined;

    set(
      fieldErrors,
      path,
      Object.assign(errors[path] || {}, { ref: field && field.ref }),
    );
  }

  return fieldErrors;
};
