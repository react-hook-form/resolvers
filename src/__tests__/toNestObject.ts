import { Field, FieldError, InternalFieldName } from 'react-hook-form';
import { toNestError } from '../toNestError';

const flatObject: Record<string, FieldError> = {
  name: { type: 'st', message: 'first message' },
  'test.0.name': { type: 'nd', message: 'second message' },
};

const fields = {
  name: {
    ref: {
      reportValidity: vi.fn(),
      setCustomValidity: vi.fn(),
    },
  },
  unused: {
    ref: { name: 'unusedRef' },
  },
} as any as Record<InternalFieldName, Field['_f']>;

test('transforms flat object to nested object', () => {
  expect(
    toNestError(flatObject, { fields, shouldUseNativeValidation: false }),
  ).toMatchSnapshot();
});

test('transforms flat object to nested object and shouldUseNativeValidation: true', () => {
  expect(
    toNestError(flatObject, { fields, shouldUseNativeValidation: true }),
  ).toMatchSnapshot();
  expect(
    (fields.name.ref as HTMLInputElement).reportValidity,
  ).toHaveBeenCalledTimes(1);
  expect(
    (fields.name.ref as HTMLInputElement).setCustomValidity,
  ).toHaveBeenCalledTimes(1);
  expect(
    (fields.name.ref as HTMLInputElement).setCustomValidity,
  ).toHaveBeenCalledWith(flatObject.name.message);
});
