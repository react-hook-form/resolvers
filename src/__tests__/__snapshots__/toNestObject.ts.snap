// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms flat object to nested object 1`] = `
{
  "name": {
    "message": "first message",
    "ref": {
      "reportValidity": [MockFunction spy],
      "setCustomValidity": [MockFunction spy],
    },
    "type": "st",
  },
  "test": [
    {
      "name": {
        "message": "second message",
        "ref": undefined,
        "type": "nd",
      },
    },
  ],
}
`;

exports[`transforms flat object to nested object and shouldUseNativeValidation: true 1`] = `
{
  "name": {
    "message": "first message",
    "ref": {
      "reportValidity": [MockFunction spy] {
        "calls": [
          [],
        ],
        "results": [
          {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "setCustomValidity": [MockFunction spy] {
        "calls": [
          [
            "first message",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    },
    "type": "st",
  },
  "test": [
    {
      "name": {
        "message": "second message",
        "ref": undefined,
        "type": "nd",
      },
    },
  ],
}
`;
